### PR TITLE
PWA: AmpDoc interface and service

### PR DESCRIPTION
**Pull request by dvoytenko**
Tests work-in-progress, but I'd like to confirm the protocol.

### CODE REVIEW

File: src/service/ampdoc-impl.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  }
+}
+
+/**
+ * @param {!Node} node
+ * @return {?ShadowRoot}
+ */
+function findShadowRoot(node) {
+  // TODO(dvoytenko): Replace with `getRootNode()` API when it's available.
+  for (let n = node; n; n = n.parentNode) {
```
Comment (cramforce) [inline src/service/ampdoc-impl.js]: Not sure this needs to be optimized? In that case, is there any chance we could use closest?

File: src/types.js [hunk 2]
```diff
@@ -69,3 +69,12 @@ export function isObject(value) {
...
   return toString(value) === '[object FormData]';
 }
+
+/**
+ * Determines if value is actually a `ShadowRoot` node.
+ * @param {*} value
+ * @return {boolean}
+ */
+export function isShadowRoot(value) {
+  return toString(value) === '[object ShadowRoot]';
```
Comment (cramforce) [inline src/types.js]: Ugh? Doesn't this have a unique node type?

Comment (dvoytenko) [inline src/types.js]: In an incredibly frustrating way (!) the node type == 11 (document fragment).

File: src/service/ampdoc-impl.js (hunk 1 above)
Comment (dvoytenko) [inline src/service/ampdoc-impl.js]: Can't find a way to optimize it via closest.

File: src/types.js (hunk 2 above)
Comment (dvoytenko) [inline src/types.js]: Added test against DOCUMENT_FRAGMENT to make the test lighter.

**Comment by dvoytenko** [comment]
@cramforce Tests completed. PTAL.

File: src/service/ampdoc-impl.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+   */
+  constructor(shadowRoot) {
+    super();
+    /** @private @const {!ShadowRoot} */
+    this.shadowRoot_ = shadowRoot;
+  }
+
+  /** @override */
+  getWin() {
+    return this.shadowRoot_.ownerDocument.defaultView;
```
Comment (cramforce) [inline src/service/ampdoc-impl.js]: Our programs are pretty sensitive to the performance of `getWin()` everywhere. Might be worth caching in the constructor.

Comment (dvoytenko) [inline src/service/ampdoc-impl.js]: Done.

**Comment by dvoytenko** [comment]
@cramforce PTAL.

**Comment by cramforce** [comment]
LGTM

**Pull request merged**